pd:  support dpa2

HydrogenSulfate opened the pull request:
Support DPA-2 in paddle backend. This PR will be updated after #4414 is merged.

### Training curve:

![training_curves_comparison_dpa2](https://github.com/user-attachments/assets/29bdeffa-cf2d-4586-afcf-7df0569997c3)



### Accuracy test(left: paddle, right: torch):

![image](https://github.com/user-attachments/assets/5bff55f3-1c39-4b95-93f0-68783e794716)


Ralated optimization of Paddle framework:
- [x] https://github.com/PaddlePaddle/Paddle/pull/69349
- [x] https://github.com/PaddlePaddle/Paddle/pull/69333
- [x] https://github.com/PaddlePaddle/Paddle/pull/69479
- [x] https://github.com/PaddlePaddle/Paddle/pull/69515
- [x] https://github.com/PaddlePaddle/Paddle/pull/69487
- [x] https://github.com/PaddlePaddle/Paddle/pull/69661
- [x] https://github.com/PaddlePaddle/Paddle/pull/69660
- [x] https://github.com/PaddlePaddle/Paddle/pull/69596
- [x] https://github.com/PaddlePaddle/Paddle/pull/69556

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced new classes for molecular descriptors: `DescrptDPA2`, `DescrptBlockRepformers`, `DescrptSeTTebd`, and `DescrptBlockSeTTebd`.
	- Added new functions for tensor operations and descriptor management, enhancing the capabilities of the module.
	- Updated JSON configurations for multitask models to refine selection criteria and data paths.

- **Bug Fixes**
	- Improved error handling and parameter validation across various descriptor classes.

- **Documentation**
	- Enhanced test coverage for new descriptor functionalities and configurations.

- **Tests**
	- Added new test classes to validate the functionality of `DescrptDPA2` and multitask training scenarios.
	- Expanded test capabilities for descriptor classes based on installed dependencies.
	- Updated existing tests to support new configurations and functionalities.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

HydrogenSulfate commented:
@njzjz This PR is ready for review

iProzd commented on deepmd/pd/model/descriptor/dpa2.py:
```diff
@@ -0,0 +1,897 @@
+        use_tebd_bias: bool = False,
+        type_map: Optional[list[str]] = None,
+    ) -> None:
+        r"""The DPA-2 descriptor. see https://arxiv.org/abs/2312.15492.
```
reference need to be updated after https://github.com/deepmodeling/deepmd-kit/pull/4483 is merged.

HydrogenSulfate replied:
> reference need to be updated after #4483 is merged.

reference now updated in `dpa2.py`.

njzjz commented:
@coderabbitai resolve

njzjz commented:
@coderabbitai resolve

njzjz commented on deepmd/pd/utils/env.py:
```diff
@@ -0,0 +1,169 @@
+    paddle.bfloat16: "bfloat16",
+    paddle.bool: "bool",
+}
+assert set(PRECISION_DICT.values()) == set(RESERVED_PRECISON_DICT.keys())
```
@HydrogenSulfate This comment seems to be reasonable

HydrogenSulfate replied:
> ```diff
> RESERVED_PRECISON_DICT 
> ```

Oh this seems a typo that exist in other backends...<https://github.com/search?q=repo%3Adeepmodeling%2Fdeepmd-kit%20RESERVED_PRECISON_DICT%20&type=code>

njzjz replied:
Maybe I can revise it in a separated PR

HydrogenSulfate replied:
> Maybe I can revise it in a separated PR

typo in pd backend are fixed now.

HydrogenSulfate commented:
@njzjz I noticed that there might be some subtle typos in the deepmd project. In Paddle, we have introduced a tool called typos in the pre-commit process, which can automatically detect potential typos. You can try it and added into deepmd if interested: <https://github.com/PaddlePaddle/Paddle/issues/69377>.
